Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs

```diff
@@ -68,8 +65,210 @@ public abstract class ResolverPass {
+        new List<AttributedExpression>(), new List<FrameExpression>(), new List<AttributedExpression>(),
+        new Specification<Expression>(new List<Expression>(), null), null, null, null) {
+        EnclosingClass = bitvectorTypeDecl,
+        ResultPreType = Type2PreType(resultType)
+      };
+      rotateMember.AddVisibilityScope(resolver.builtIns.SystemModule.VisibilityScope, false);
+      bitvectorTypeDecl.Members.Add(name, rotateMember);
+    }
+
+    TopLevelDecl BuiltInArrowTypeDecl(int arity) {
+      Contract.Requires(0 <= arity);
+      var name = ArrowType.ArrowTypeName(arity);
+      if (!preTypeBuiltins.TryGetValue(name, out var decl)) {
+        // the arrow type declaration should already have been created by the parser
+        decl = resolver.builtIns.ArrowTypeDecls[arity];
+        preTypeBuiltins.Add(name, decl);
+      }
+      return decl;
+    }
+
+    DPreType BuiltInArrowType(List<PreType> inPreTypes, PreType resultPreType) {
+      return new DPreType(BuiltInArrowTypeDecl(inPreTypes.Count), Util.Snoc(inPreTypes, resultPreType));
+    }
+
+    DPreType BuiltInArrayType(int dims, PreType elementPreType) {
+      Contract.Requires(1 <= dims);
+      var arrayName = dims == 1 ? "array" : $"array{dims}";
+      return new DPreType(BuiltInTypeDecl(arrayName), new List<PreType>() { elementPreType });
+    }
+
+    private int typeProxyCount = 0; // used to give each PreTypeProxy a unique ID
+
+    private readonly List<(PreTypeProxy, string)> allPreTypeProxies = new();
+
+    public PreType CreatePreTypeProxy(string description = null) {
+      var proxy = new PreTypeProxy(typeProxyCount++);
+      allPreTypeProxies.Add((proxy, description));
+      return proxy;
+    }
+
+    public enum Type2PreTypeOption { GoodForInference, GoodForPrinting, GoodForBoth }
+
+    public PreType Type2PreType(Type type, string description = null, Type2PreTypeOption option = Type2PreTypeOption.GoodForBoth) {
+      Contract.Requires(type != null);
+
+      type = type.Normalize();
+      var expandedType = type.NormalizeExpand();
+      if (expandedType is TypeProxy) {
+        return CreatePreTypeProxy(description ?? $"from type proxy {type}");
+      }
+
+      DPreType printablePreType = null;
+      if (option != Type2PreTypeOption.GoodForInference) {
+        var printableDecl = Type2PreTypeDecl(type);
+        var printableArguments = type.TypeArgs.ConvertAll(ty => Type2PreType(ty, null, Type2PreTypeOption.GoodForPrinting));
+        printablePreType = new DPreType(printableDecl, printableArguments, null);
+        if (option == Type2PreTypeOption.GoodForPrinting) {
+          return printablePreType;
+        }
+      }
+
+      type = expandedType;
+      var decl = Type2PreTypeDecl(type);
+      var arguments = type.TypeArgs.ConvertAll(ty => Type2PreType(ty, null, Type2PreTypeOption.GoodForInference));
+      return new DPreType(decl, arguments, printablePreType);
+    }
+
+    TopLevelDecl Type2PreTypeDecl(Type type) {
```

Comment:
Given the return type, could perhaps be renamed more generally to  `Type2Decl`. Will there be a `Type2DPreType` function, too?

Reply:
I like it. I'll rename it to `Type2Decl`.